aya/src/programs/uprobe.rs

```diff
@@ -206,66 +206,119 @@ pub(crate) struct LdSoCache {
     entries: Vec<CacheEntry>,
 }
 
+#[allow(unused)]
+#[derive(Copy, Clone, Debug)]
+pub enum TargetEndian {
+    Native,
+    Big,
+    Little,
+}
+
 impl LdSoCache {
     pub fn load<T: AsRef<Path>>(path: T) -> Result<Self, io::Error> {
         let data = fs::read(path)?;
-        Self::parse(&data)
+        Self::parse(&data, TargetEndian::Native)
     }
 
-    fn parse(data: &[u8]) -> Result<Self, io::Error> {
+    fn parse(data: &[u8], endianness: TargetEndian) -> Result<Self, io::Error> {
         let mut cursor = Cursor::new(data);
 
         let read_u32 = |cursor: &mut Cursor<_>| -> Result<u32, io::Error> {
             let mut buf = [0u8; mem::size_of::<u32>()];
             cursor.read_exact(&mut buf)?;
 
-            Ok(u32::from_ne_bytes(buf))
+            Ok(match endianness {
+                TargetEndian::Native => u32::from_ne_bytes(buf),
+                TargetEndian::Big => u32::from_be_bytes(buf),
+                TargetEndian::Little => u32::from_le_bytes(buf),
+            })
         };
 
         let read_i32 = |cursor: &mut Cursor<_>| -> Result<i32, io::Error> {
             let mut buf = [0u8; mem::size_of::<i32>()];
             cursor.read_exact(&mut buf)?;
-
-            Ok(i32::from_ne_bytes(buf))
+            Ok(match endianness {
+                TargetEndian::Native => i32::from_ne_bytes(buf),
+                TargetEndian::Big => i32::from_be_bytes(buf),
+                TargetEndian::Little => i32::from_le_bytes(buf),
+            })
         };
 
         let mut buf = [0u8; LD_SO_CACHE_HEADER.len()];
+        let mut buf_old = [0u8; LD_SO_CACHE_HEADER_OLD.len()];
         cursor.read_exact(&mut buf)?;
-        let header = std::str::from_utf8(&buf).or(Err(io::Error::new(
-            io::ErrorKind::InvalidData,
-            "invalid ld.so.cache header",
-        )))?;
-        if header != LD_SO_CACHE_HEADER {
-            return Err(io::Error::new(
-                io::ErrorKind::InvalidData,
-                "invalid ld.so.cache header",
-            ));
+        cursor.set_position(0);
+        cursor.read_exact(&mut buf_old)?;
+        let header = std::str::from_utf8(&buf).or(
```

Comment:
See my comment here : https://github.com/EKTehtris/ld_cache_rs/blob/main/src/lib.rs#L374

Reply:
However it's rarely (and to be exact never the case because of the format of the separator that it's valid utf8)

Reply:
I'll see if I can reproduce it (though I don't remember off hand where I ran into it), but I'm pretty sure I ran into a case where the old header with trailing content parsed as valid utf-8. What's the reason for parsing to utf-8 anyway instead of a direct bytes comparison? I don't see what the benefit is here.

Reply:
This was to avoid the (im)possible case where you would have multibytes encoding that would match the header at the bytes level but not be the same when decoded as utf8.